File: java/structures/Treap.java
package structures;

import java.util.Random;

// https://cp-algorithms.com/data_structures/treap.html
public class Treap {
    static Random random = new Random();

    public static class Node {
        long nodeValue;
        long mx;
        long sum;
        long add;

        long key; // keys should be unique
        int size;
        long prio;
        Node left;
        Node right;

        Node(long key, long value) {
            this.key = key;
            nodeValue = value;
            mx = value;
            sum = value;
            add = 0;
            size = 1;
            prio = random.nextLong();
        }

        void apply(long v) {
            nodeValue += v;
            mx += v;
            sum += v * size;
            add += v;
        }

        void push() {
            if (add != 0) {
                if (left != null)
                    left.apply(add);
                if (right != null)
                    right.apply(add);
                add = 0;
            }
        }

        void pull() {
            mx = Math.max(nodeValue, Math.max(getMx(left), getMx(right)));
            sum = nodeValue + getSum(left) + getSum(right);
            size = 1 + getSize(left) + getSize(right);
        }

        static long getMx(Node root) {
            return root == null ? Long.MIN_VALUE : root.mx;
        }

        static long getSum(Node root) {
            return root == null ? 0 : root.sum;
        }

        static int getSize(Node root) {
            return root == null ? 0 : root.size;
        }
    }

    public static class TreapPair {
        Node left;
        Node right;

        TreapPair(Node left, Node right) {
            this.left = left;
            this.right = right;
        }
    }

    public static TreapPair split(Node root, long minRight) {
        if (root == null)
            return new TreapPair(null, null);
        root.push();
        if (root.key >= minRight) {
            TreapPair sub = split(root.left, minRight);
            root.left = sub.right;
            root.pull();
            sub.right = root;
            return sub;
        } else {
            TreapPair sub = split(root.right, minRight);
            root.right = sub.left;
            root.pull();
            sub.left = root;
            return sub;
        }
    }

    public static Node merge(Node left, Node right) {
        if (left == null)
            return right;
        if (right == null)
            return left;
        left.push();
        right.push();
        if (left.prio > right.prio) {
            left.right = merge(left.right, right);
            left.pull();
            return left;
        } else {
            right.left = merge(left, right.left);
            right.pull();
            return right;
        }
    }

    public static Node insert(Node root, long key, long value) {
        TreapPair t = split(root, key);
        return merge(merge(t.left, new Node(key, value)), t.right);
    }

    public static Node remove(Node root, long key) {
        TreapPair t = split(root, key);
        return merge(t.left, split(t.right, key + 1).right);
    }

    public static Node modify(Node root, long ll, long rr, long delta) {
        TreapPair t1 = split(root, rr + 1);
        TreapPair t2 = split(t1.left, ll);

        if (t2.right != null)
            t2.right.apply(delta);
        return merge(merge(t2.left, t2.right), t1.right);
    }

    public static class TreapAndResult {
        Node treap;
        long mx;
        long sum;

        TreapAndResult(Node t, long mx, long sum) {
            this.treap = t;
            this.mx = mx;
            this.sum = sum;
        }
    }

    public static TreapAndResult query(Node root, long ll, long rr) {
        TreapPair t1 = split(root, rr + 1);
        TreapPair t2 = split(t1.left, ll);
        long mx = Node.getMx(t2.right);
        long sum = Node.getSum(t2.right);
        return new TreapAndResult(merge(merge(t2.left, t2.right), t1.right), mx, sum);
    }

    static Node kth(Node root, int k) {
        if (k < Node.getSize(root.left))
            return kth(root.left, k);
        else if (k > Node.getSize(root.left))
            return kth(root.right, k - Node.getSize(root.left) - 1);
        return root;
    }

    public static void print(Node root) {
        if (root == null)
            return;
        root.push();
        print(root.left);
        System.out.print(root.nodeValue + " ");
        print(root.right);
    }

    // Random test
    public static void main(String[] args) {
        Node treap = null;
        treap = insert(treap, 5, 3);
        treap = insert(treap, 3, 2);
        treap = insert(treap, 6, 1);
        System.out.println(kth(treap, 1).key);
        System.out.println(query(treap, 1, 10).mx);
        treap = remove(treap, 5);
        System.out.println(query(treap, 1, 10).mx);
    }
}

File: rust/structures/treap.rs
use rand::Rng;
use std::cmp::max;
use std::ops::AddAssign;

struct Treap<K, V> {
    value: V,
    mx: V,
    add: V,
    // keys should be unique
    key: K,
    prio: i64,
    size: usize,
    left: Option<Box<Treap<K, V>>>,
    right: Option<Box<Treap<K, V>>>,
}

impl<K: Copy + PartialOrd, V: Copy + PartialEq + AddAssign + Ord + Default> Treap<K, V> {
    fn new(key: K, value: V) -> Option<Box<Treap<K, V>>> {
        Some(Box::new(Self {
            key,
            value,
            mx: value,
            add: V::default(),
            prio: rand::thread_rng().gen_range(0..1_000_000_000),
            size: 1,
            left: None,
            right: None,
        }))
    }

    fn apply(&mut self, v: V) {
        self.value += v;
        self.mx += v;
        self.add += v;
    }

    fn push(&mut self) {
        if self.add != V::default() {
            if self.left.is_some() {
                self.left.as_mut().unwrap().apply(self.add);
            }
            if self.right.is_some() {
                self.right.as_mut().unwrap().apply(self.add);
            }
            self.add = V::default();
        }
    }

    fn pull(&mut self) {
        self.size = 1 + Self::size(&self.left) + Self::size(&self.right);
        self.mx = Self::mx(Self::mx(self.value, &self.left), &self.right);
    }

    pub fn size(node: &Option<Box<Treap<K, V>>>) -> usize {
        node.as_ref().map_or(0, |t| t.size)
    }

    fn mx(value: V, node: &Option<Box<Treap<K, V>>>) -> V {
        node.as_ref().map_or(value, |t| max(t.mx, value))
    }

    #[allow(clippy::type_complexity)]
    pub fn split(
        root: Option<Box<Treap<K, V>>>,
        min_right: K,
    ) -> (Option<Box<Treap<K, V>>>, Option<Box<Treap<K, V>>>) {
        Self::inner_split(root, min_right, |a, b| a >= b)
    }

    #[allow(clippy::type_complexity)]
    pub fn strict_split(
        root: Option<Box<Treap<K, V>>>,
        min_right: K,
    ) -> (Option<Box<Treap<K, V>>>, Option<Box<Treap<K, V>>>) {
        Self::inner_split(root, min_right, |a, b| a > b)
    }

    #[allow(clippy::type_complexity)]
    fn inner_split<F: Fn(K, K) -> bool>(
        root: Option<Box<Treap<K, V>>>,
        min_right: K,
        cmp: F,
    ) -> (Option<Box<Treap<K, V>>>, Option<Box<Treap<K, V>>>) {
        if root.is_none() {
            return (None, None);
        }
        let mut node = root.unwrap();
        node.push();
        if cmp(node.key, min_right) {
            let (l, r) = Self::inner_split(node.left, min_right, cmp);
            node.left = r;
            node.pull();
            (l, Some(node))
        } else {
            let (l, r) = Self::inner_split(node.right, min_right, cmp);
            node.right = l;
            node.pull();
            (Some(node), r)
        }
    }

    pub fn merge(
        a: Option<Box<Treap<K, V>>>,
        b: Option<Box<Treap<K, V>>>,
    ) -> Option<Box<Treap<K, V>>> {
        if a.is_none() {
            return b;
        }
        if b.is_none() {
            return a;
        }
        let mut left = a.unwrap();
        let mut right = b.unwrap();
        left.push();
        right.push();
        if left.prio > right.prio {
            left.right = Self::merge(left.right, Some(right));
            left.pull();
            Some(left)
        } else {
            right.left = Self::merge(Some(left), right.left);
            right.pull();
            Some(right)
        }
    }

    pub fn insert(root: Option<Box<Treap<K, V>>>, key: K, value: V) -> Option<Box<Treap<K, V>>> {
        let (l, r) = Self::split(root, key);
        Self::merge(Self::merge(l, Treap::new(key, value)), r)
    }

    pub fn remove(root: Option<Box<Treap<K, V>>>, key: K) -> Option<Box<Treap<K, V>>> {
        let (l, r) = Self::split(root, key);
        Self::merge(l, Self::strict_split(r, key).1)
    }

    pub fn modify(
        root: Option<Box<Treap<K, V>>>,
        ll: K,
        rr: K,
        delta: V,
    ) -> Option<Box<Treap<K, V>>> {
        let (l1, r1) = Self::strict_split(root, rr);
        let (l2, mut r2) = Self::split(l1, ll);
        #[allow(clippy::option_map_unit_fn)]
        r2.as_mut().map(|t| t.apply(delta));
        Self::merge(Self::merge(l2, r2), r1)
    }

    pub fn query(root: Option<Box<Treap<K, V>>>, ll: K, rr: K) -> (Option<Box<Treap<K, V>>>, V) {
        let (l1, r1) = Self::strict_split(root, rr);
        let (l2, r2) = Self::split(l1, ll);
        let mx = r2.as_ref().map_or(V::default(), |t| t.mx);
        (Self::merge(Self::merge(l2, r2), r1), mx)
    }
}

#[cfg(test)]
mod tests {
    use crate::structures::treap::Treap;

    #[test]
    fn basic_test() {
        let mut treap: Option<Box<Treap<u32, u32>>> = None;
        treap = Treap::insert(treap, 1, 10);
        treap = Treap::insert(treap, 2, 20);
        let (t, q1) = Treap::query(treap, 1, 2);
        treap = t;
        assert_eq!(q1, 20);

        treap = Treap::modify(treap, 1, 2, 100);

        let (t, q2) = Treap::query(treap, 1, 2);
        treap = t;
        assert_eq!(q2, 120);

        treap = Treap::remove(treap, 2);

        let (_t, q3) = Treap::query(treap, 1, 2);
        assert_eq!(q3, 110);
    }
}
